Agroal configuration rewrite

gsmet opened the pull request:
This is the PR that rewrites the Agroal configuration to allow for runtime configuration.

It also required some changes in the Hibernate extension to take into account the fact that datasources are initialized later.

There is still quite a lot of work to do to support multiple datasources cleanly (the Agroal part is ready pending issues in Config but the Hibernate extension is not ready yet) but we decided it was not a priority for first public release so I will do that later.

In the ORM extension, I also tried to make things more readable, avoiding passing around a mutable `Map` for the Hibernate Configuration.

Let's see what CI has to say as it will use the new runtime config to pass the datasource url.

gsmet committed:
Agroal - Configuration rewrite

We aim to support multiple datasources at some point but we are not
there yet.

gsmet committed:
Agroal - Only test the default config for now

gsmet committed:
Agroal - Add some useful configuration knobs

gsmet committed:
Push the datasource URL through Shamrock config

We don't support environment variables anymore.

gsmet committed:
Agroal - Throw an exception if runtime configuration not present yet

gsmet committed:
Agroal - Make everything public due to a bug in Gizmo

gsmet committed:
Agroal - Mark the datasource bean as unremovable

gsmet committed:
Hibernate ORM - Only starts persistence units when datasources are fully configured

gsmet committed:
Hibernate ORM - Properly separate build time and runtime settings

We are talking about the Map of settings we push to Hibernate, not the
Quarkus config yet.

gsmet committed:
Hibernate ORM - Reduce the number of warnings

gsmet committed:
Hibernate ORM - Initialize the datasource later

gsmet committed:
Agroal - Always use JTA

Sanne reviewed (commented):
Great progress, thanks! Had some very minor observations, feel free to choose which one you'd like to fix or ignore

gsmet committed:
Agroal - Consider the MariaDB driver when disabling SSL

This error was preexisting but it needs fixing: we support the MariaDB
driver, not the MySQL one.

gsmet committed:
Agroal - Add a comment about the purpose of the DataSourceProducer bean

gsmet committed:
Hibernate ORM - Throw an exception if datasource provided in persistence units

Sanne reviewed (approved):
Explicitly approving so that you can push whenever you're done.
